pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/rebalance/DefaultRebalanceSegmentStrategy.java

```diff
@@ -161,20 +154,55 @@ public IdealState rebalanceIdealState(IdealState idealState, TableConfig tableCo
       targetNumReplicas = Integer.parseInt(tableConfig.getValidationConfig().getReplication());
     }
 
-    // always rebalance serving segments
-    rebalanceServingSegments(idealState, tableConfig, targetNumReplicas);
-
     // update if not dryRun
     boolean dryRun = rebalanceUserConfig.getBoolean(RebalanceUserConfigConstants.DRYRUN, DEFAULT_DRY_RUN);
     if (!dryRun) {
       LOGGER.info("Updating ideal state for table {}", tableNameWithType);
-      updateIdealState(tableNameWithType, targetNumReplicas, idealState.getRecord().getMapFields());
+      rebalanceAndUpdateIdealState(idealState, tableConfig, targetNumReplicas, rebalanceUserConfig,
+          newPartitionAssignment);
     } else {
       LOGGER.info("Dry run. Skip writing ideal state");
+      rebalanceIdealState(idealState, tableConfig, targetNumReplicas, rebalanceUserConfig, newPartitionAssignment);
     }
     return idealState;
   }
 
+  private void rebalanceAndUpdateIdealState(final IdealState rebalanceIdealState, final TableConfig tableConfig,
+      final int targetNumReplicas, final Configuration rebalanceUserConfig,
+      final PartitionAssignment newPartitionAssignment) {
+
+    final Function<IdealState, IdealState> updaterFunction = new Function<IdealState, IdealState>() {
+      @Nullable
+      @Override
+      public IdealState apply(@Nullable IdealState idealState) {
+        rebalanceIdealState(rebalanceIdealState, tableConfig, targetNumReplicas, rebalanceUserConfig,
+            newPartitionAssignment);
+        Map<String, Map<String, String>> mapFields = rebalanceIdealState.getRecord().getMapFields();
```

Comment:
fold these lines into rebalanceIdelState if it is not there already

Reply:
removed ideal state argument and removed the lines to copy over map